File: foms/services/designer/formula_engine.py
"""FOMS Brain Design Kernel V1 — Formula Engine (Python backend).

DK-B2: deterministic formula evaluator.

Features:
- assembly/module/component dimension 참조
- parent dimension 변경 시 child 재계산
- mm 정수 normalize
- circular formula 감지
"""

from __future__ import annotations

import math
from typing import Any

from foms.services.designer.ontology_types import DesignGraph, Component

# ──────────────────────────────────────────────────────────
# Formula variable resolution
# ──────────────────────────────────────────────────────────

_SAFE_MATH = {
    "abs": abs,
    "round": round,
    "floor": math.floor,
    "ceil": math.ceil,
    "int": int,
}


def _build_context(graph: DesignGraph, component: Component | None = None) -> dict[str, Any]:
    """Build formula evaluation context from design graph."""
    asm = graph.assembly
    ctx: dict[str, Any] = {
        # assembly-level shortcuts
        "outer_width": asm.dimensions.width,
        "outer_height": asm.dimensions.height,
        "outer_depth": asm.dimensions.depth,
        "total_width": asm.dimensions.width,
        "total_height": asm.dimensions.height,
        "total_depth": asm.dimensions.depth,
        "ep_left": asm.ep_left,
        "ep_right": asm.ep_right,
        "ep_top": asm.ep_top,
        "base": asm.base_height,
        "top_sr": asm.top_sr,
        "module_count": asm.module_count,
        # common gaps / constants
        "gap": 2,      # 도어 간격 기본값 2mm
        "pb_t": 18,    # PB 기본 두께
        "back_t": 9,   # 후판 두께
    }

    # assembly namespace
    ctx["assembly"] = {
        "width": asm.dimensions.width,
        "height": asm.dimensions.height,
        "depth": asm.dimensions.depth,
        "ep_left": asm.ep_left,
        "ep_right": asm.ep_right,
        "ep_top": asm.ep_top,
        "base_height": asm.base_height,
        "top_sr": asm.top_sr,
        "module_count": asm.module_count,
    }

    if component:
        ctx["this"] = {
            "width": component.dimensions.width,
            "height": component.dimensions.height,
            "depth": component.dimensions.depth,
        }
        # parent (module) dimension lookup
        if component.parent_id:
            parent_mod = graph.get_module(component.parent_id)
            if parent_mod:
                ctx["parent"] = {
                    "width": parent_mod.dimensions.width,
                    "height": parent_mod.dimensions.height,
                    "depth": parent_mod.dimensions.depth,
                }

    return ctx


def _normalize_mm(value: float) -> int:
    """Round formula result to nearest mm integer."""
    return int(round(value))


def _eval_expr(expression: str, context: dict[str, Any]) -> int:
    """
    Safely evaluate a formula expression string.
    Supports only arithmetic + context variables.
    """
    try:
        result = eval(  # noqa: S307
            expression,
            {"__builtins__": {}},
            {**_SAFE_MATH, **context},
        )
        if not isinstance(result, (int, float)):
            raise ValueError(f"공식 결과가 숫자가 아닙니다: {result!r}")
        return _normalize_mm(result)
    except Exception as exc:
        raise FormulaError(f"공식 평가 실패: {expression!r} — {exc}") from exc


# ──────────────────────────────────────────────────────────
# Built-in formula library
# ──────────────────────────────────────────────────────────

BUILTIN_FORMULAS: dict[str, dict[str, Any]] = {
    "module_width": {
        "expression": "(outer_width - ep_left - ep_right) / module_count",
        "description": "모듈 폭 = (전체 폭 - 좌EP - 우EP) / 모듈수",
    },
    "door_height": {
        "expression": "outer_height - top_sr - base - gap",
        "description": "도어 높이 = 전체 높이 - 상부SR - 받침대 - 간격",
    },
    "inner_height": {
        "expression": "outer_height - top_sr - base - pb_t * 2",
        "description": "내부 유효 높이 = 전체 높이 - SR - 받침대 - 상하판",
    },
    "inner_width_per_module": {
        "expression": "(outer_width - ep_left - ep_right - pb_t * (module_count - 1)) / module_count",
        "description": "모듈 내부 폭 (칸막이 판재 포함 공제)",
    },
    "back_panel_width": {
        "expression": "outer_width",
        "description": "후판 폭 = 전체 폭",
    },
    "back_panel_height": {
        "expression": "outer_height",
        "description": "후판 높이 = 전체 높이",
    },
    "side_panel_height": {
        "expression": "outer_height",
        "description": "측판 높이 = 전체 높이",
    },
    "shelf_width": {
        "expression": "parent.width - pb_t * 2",
        "description": "선반 폭 = 모듈 폭 - 좌우 판재 두께",
    },
}


def evaluate_formula(
    formula_id: str,
    graph: DesignGraph,
    component: Component | None = None,
    overrides: dict[str, Any] | None = None,
) -> int:
    """Evaluate a built-in formula by id against the design graph.

    Args:
        formula_id: key in BUILTIN_FORMULAS
        graph: current DesignGraph
        component: optional component context (for 'parent'/'this' refs)
        overrides: extra variable overrides

    Returns:
        mm integer result

    Raises:
        FormulaError on unknown formula or evaluation failure.
    """
    if formula_id not in BUILTIN_FORMULAS:
        raise FormulaError(f"알 수 없는 공식: {formula_id!r}")

    formula = BUILTIN_FORMULAS[formula_id]
    ctx = _build_context(graph, component)
    if overrides:
        ctx.update(overrides)
    return _eval_expr(formula["expression"], ctx)


def evaluate_expression(
    expression: str,
    graph: DesignGraph,
    component: Component | None = None,
    overrides: dict[str, Any] | None = None,
) -> int:
    """Evaluate an arbitrary formula expression string against the design graph."""
    ctx = _build_context(graph, component)
    if overrides:
        ctx.update(overrides)
    return _eval_expr(expression, ctx)


# ──────────────────────────────────────────────────────────
# Circular dependency detection
# ──────────────────────────────────────────────────────────

def check_circular_formulas(formula_deps: dict[str, list[str]]) -> list[str]:
    """
    Detect circular dependencies in a formula dependency graph.

    Args:
        formula_deps: {formula_id: [dependent_formula_ids]}

    Returns:
        List of formula ids that form cycles (empty if none).
    """
    visited: set[str] = set()
    in_stack: set[str] = set()
    cycles: list[str] = []

    def dfs(node: str) -> bool:
        visited.add(node)
        in_stack.add(node)
        for dep in formula_deps.get(node, []):
            if dep not in visited:
                if dfs(dep):
                    cycles.append(node)
                    return True
            elif dep in in_stack:
                cycles.append(node)
                return True
        in_stack.discard(node)
        return False

    for fid in formula_deps:
        if fid not in visited:
            dfs(fid)

    return cycles


# ──────────────────────────────────────────────────────────
# Batch recalculate: apply formulas to all components in graph
# ──────────────────────────────────────────────────────────

def recalculate_graph(graph: DesignGraph) -> list[dict]:
    """
    Recalculate all components that have formula_refs.
    Returns list of {component_id, prop, before, after} dicts for audit.
    """
    changes: list[dict] = []

    for comp in graph.components:
        for formula_id in comp.formula_refs:
            if formula_id not in BUILTIN_FORMULAS:
                continue
            try:
                new_value = evaluate_formula(formula_id, graph, comp)
                # Determine which dimension this formula targets
                target = BUILTIN_FORMULAS[formula_id].get("description", "")
                if "높이" in target or "height" in formula_id:
                    old = comp.dimensions.height
                    if old != new_value:
                        changes.append({
                            "component_id": comp.id,
                            "prop": "dimensions.height",
                            "before": old,
                            "after": new_value,
                        })
                        comp.dimensions.height = new_value
                elif "폭" in target or "width" in formula_id:
                    old = comp.dimensions.width
                    if old != new_value:
                        changes.append({
                            "component_id": comp.id,
                            "prop": "dimensions.width",
                            "before": old,
                            "after": new_value,
                        })
                        comp.dimensions.width = new_value
            except FormulaError:
                pass  # non-blocking; constraint engine will catch invalid states

    return changes


# ──────────────────────────────────────────────────────────
# Error
# ──────────────────────────────────────────────────────────

class FormulaError(Exception):
    """Raised when a formula cannot be evaluated."""

File: Add In Program/FOMSBrainDesigner/src/domain/formulaEngine.ts
/**
 * FOMS Brain Design Kernel V1 — Formula Engine (TypeScript frontend)
 * DK-B2: deterministic formula evaluator
 *
 * Features:
 * - assembly/module/component dimension 참조
 * - parent dimension 변경 시 child 재계산
 * - mm 정수 normalize
 * - circular formula 감지
 */

import type { DesignGraph, Component, Assembly } from './ontologyTypes'

// ──────────────────────────────────────────────────────────
// Built-in formula library
// ──────────────────────────────────────────────────────────

export interface BuiltinFormula {
  id: string
  expression: string
  description: string
  targets: 'width' | 'height' | 'depth'
}

export const BUILTIN_FORMULAS: Record<string, BuiltinFormula> = {
  module_width: {
    id: 'module_width',
    expression: '(outer_width - ep_left - ep_right) / module_count',
    description: '모듈 폭 = (전체 폭 - 좌EP - 우EP) / 모듈수',
    targets: 'width',
  },
  door_height: {
    id: 'door_height',
    expression: 'outer_height - top_sr - base - gap',
    description: '도어 높이 = 전체 높이 - 상부SR - 받침대 - 간격',
    targets: 'height',
  },
  inner_height: {
    id: 'inner_height',
    expression: 'outer_height - top_sr - base - pb_t * 2',
    description: '내부 유효 높이',
    targets: 'height',
  },
  inner_width_per_module: {
    id: 'inner_width_per_module',
    expression: '(outer_width - ep_left - ep_right - pb_t * (module_count - 1)) / module_count',
    description: '모듈 내부 폭',
    targets: 'width',
  },
  back_panel_width: {
    id: 'back_panel_width',
    expression: 'outer_width',
    description: '후판 폭',
    targets: 'width',
  },
  back_panel_height: {
    id: 'back_panel_height',
    expression: 'outer_height',
    description: '후판 높이',
    targets: 'height',
  },
  side_panel_height: {
    id: 'side_panel_height',
    expression: 'outer_height',
    description: '측판 높이',
    targets: 'height',
  },
  shelf_width: {
    id: 'shelf_width',
    expression: 'parent_width - pb_t * 2',
    description: '선반 폭 = 모듈 폭 - 좌우 판재 두께',
    targets: 'width',
  },
}

// ──────────────────────────────────────────────────────────
// Context builder
// ──────────────────────────────────────────────────────────

export interface FormulaContext {
  outer_width: number
  outer_height: number
  outer_depth: number
  total_width: number
  total_height: number
  total_depth: number
  ep_left: number
  ep_right: number
  ep_top: number
  base: number
  top_sr: number
  module_count: number
  gap: number
  pb_t: number
  back_t: number
  parent_width?: number
  parent_height?: number
  parent_depth?: number
  [key: string]: number | undefined
}

export function buildContext(graph: DesignGraph, component?: Component): FormulaContext {
  const asm = graph.assembly
  const ctx: FormulaContext = {
    outer_width: asm.dimensions.width,
    outer_height: asm.dimensions.height,
    outer_depth: asm.dimensions.depth,
    total_width: asm.dimensions.width,
    total_height: asm.dimensions.height,
    total_depth: asm.dimensions.depth,
    ep_left: asm.ep_left,
    ep_right: asm.ep_right,
    ep_top: asm.ep_top,
    base: asm.base_height,
    top_sr: asm.top_sr,
    module_count: asm.module_count,
    gap: 2,
    pb_t: 18,
    back_t: 9,
  }

  if (component?.parent_id) {
    const parentModule = asm.modules.find(m => m.id === component.parent_id)
    if (parentModule) {
      ctx.parent_width = parentModule.dimensions.width
      ctx.parent_height = parentModule.dimensions.height
      ctx.parent_depth = parentModule.dimensions.depth
    }
  }

  return ctx
}

// ──────────────────────────────────────────────────────────
// Expression evaluator (safe subset)
// ──────────────────────────────────────────────────────────

export function evaluateExpression(expression: string, context: FormulaContext): number {
  // Safe evaluation using Function with explicit variable injection
  const varNames = Object.keys(context)
  const varValues = varNames.map(k => context[k] ?? 0)

  try {
    // eslint-disable-next-line no-new-func
    const fn = new Function(...varNames, `"use strict"; return (${expression});`)
    const result = fn(...varValues)
    if (typeof result !== 'number' || !isFinite(result)) {
      throw new FormulaError(`공식 결과가 유효한 숫자가 아닙니다: ${result}`)
    }
    return normalizeMm(result)
  } catch (e) {
    if (e instanceof FormulaError) throw e
    throw new FormulaError(`공식 평가 실패: ${expression} — ${e}`)
  }
}

export function normalizeMm(value: number): number {
  return Math.round(value)
}

// ──────────────────────────────────────────────────────────
// Built-in formula evaluation
// ──────────────────────────────────────────────────────────

export function evaluateFormula(
  formulaId: string,
  graph: DesignGraph,
  component?: Component,
  overrides?: Partial<FormulaContext>,
): number {
  const formula = BUILTIN_FORMULAS[formulaId]
  if (!formula) throw new FormulaError(`알 수 없는 공식: ${formulaId}`)

  const ctx = buildContext(graph, component)
  if (overrides) Object.assign(ctx, overrides)
  return evaluateExpression(formula.expression, ctx)
}

// ──────────────────────────────────────────────────────────
// Circular dependency detection
// ──────────────────────────────────────────────────────────

export function checkCircularFormulas(formulaDeps: Record<string, string[]>): string[] {
  const visited = new Set<string>()
  const inStack = new Set<string>()
  const cycles: string[] = []

  function dfs(node: string): boolean {
    visited.add(node)
    inStack.add(node)
    for (const dep of formulaDeps[node] ?? []) {
      if (!visited.has(dep)) {
        if (dfs(dep)) { cycles.push(node); return true }
      } else if (inStack.has(dep)) {
        cycles.push(node)
        return true
      }
    }
    inStack.delete(node)
    return false
  }

  for (const fid of Object.keys(formulaDeps)) {
    if (!visited.has(fid)) dfs(fid)
  }
  return cycles
}

// ──────────────────────────────────────────────────────────
// Batch recalculate
// ──────────────────────────────────────────────────────────

export interface FormulaChange {
  componentId: string
  prop: 'width' | 'height' | 'depth'
  before: number
  after: number
}

export function recalculateGraph(graph: DesignGraph): { graph: DesignGraph; changes: FormulaChange[] } {
  const changes: FormulaChange[] = []
  const components = graph.components.map(comp => ({ ...comp, dimensions: { ...comp.dimensions } }))

  for (const comp of components) {
    for (const formulaId of comp.formula_refs ?? []) {
      const formula = BUILTIN_FORMULAS[formulaId]
      if (!formula) continue
      try {
        const newValue = evaluateFormula(formulaId, graph, comp)
        const target = formula.targets
        const oldValue = comp.dimensions[target]
        if (oldValue !== newValue) {
          changes.push({ componentId: comp.id, prop: target, before: oldValue, after: newValue })
          comp.dimensions[target] = newValue
        }
      } catch {
        // Non-blocking; constraint engine will catch invalid states
      }
    }
  }

  return { graph: { ...graph, components }, changes }
}

// ──────────────────────────────────────────────────────────
// Error
// ──────────────────────────────────────────────────────────

export class FormulaError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FormulaError'
  }
}
